STM32/AC/HeatCtrl/Src/HeatCtrl.c

```diff
@@ -0,0 +1,148 @@
+#include "HeatCtrl.h"
+
+#define MAX_DURATION ((uint32_t) -1)
+
+typedef struct HeatCtrl
+{
+    // pwm data.
+    uint32_t pwmDuration;
+    uint32_t pwmPeriod;  // Period (in mSec) for a full PWM signal
+    uint8_t  pwmPercent; // Percentage of the PWM signal where is should be high.
+
+    StmGpio *heater;
+    StmGpio *button;
+
+    // timeStamp begin of period. Used to safety shut off and PWM signal generation
+    uint32_t periodBegin;
+} HeatCtrl;
+
+static HeatCtrl heaters[MAX_NO_HEATERS];
+static int noOfHeaters = 0;
+
+HeatCtrl* heatCtrlAdd(StmGpio *heater, StmGpio * button)
+{
+    if (noOfHeaters >= MAX_NO_HEATERS)
+        return NULL;
+
+    HeatCtrl *ctx = &heaters[noOfHeaters];
+    noOfHeaters++;
+
+    ctx->pwmDuration = 0;     // Default is off.
+    ctx->pwmPeriod   = 1000;  // default value, 1 seconds.
+    ctx->pwmPercent  = 0;     // Default is off.
+
+    ctx->heater = heater;
+    ctx->button = button;
+
+    ctx->periodBegin = HAL_GetTick();
+    return ctx;
+}
+
+// Finds the time difference head-tail and assume head >= tail.
+// If head < tail it is assumed that head is overflow.
+static uint32_t tDiff(uint32_t head, uint32_t tail)
+{
+    if (head >= tail)
+        return head - tail;
+
+    // Overflow of head:
+    // (head + 2^32) - tail = (2^32 - tail) + head = (see below)
+    return (((uint32_t) -1) - tail) + 1 + head;
+}
+
```

Comment:
If I remember correct then this method is the only one that changes heater output. This is "critical" to your design. I would add that in a comment.